### PR TITLE
[F] Add draft mode to projects

### PR DESCRIPTION
**Pull request by SMaxOwok**

### CODE REVIEW

**Comment by zdavis** [comment]
@SMaxOwok please rebase onto master when you have a chance.

**Comment by SMaxOwok** [comment]
@zdavis rebased on master

File: api/app/controllers/api/v1/projects_controller.rb [hunk 1]
```diff
@@ -27,6 +27,7 @@ def index
       def show
         @scope_for_projects = Project.friendly
         @project = load_project
+        authorize_action_for @project if @project.draft
```
Comment (zdavis) [inline api/app/controllers/api/v1/projects_controller.rb]: Can you help me understand what's happening here? This doesn't make sense to me. It seems to me that we should always authorize the project show method and handle any conditional logic in the project authorizer. If the project is set to draft and the user is not an admin, then the user is NOT authorized to show the project.

File: api/app/models/project.rb [hunk 2]
```diff
@@ -82,6 +82,7 @@ class Project < ApplicationRecord
 
   # Callbacks
   after_commit :trigger_creation_event, on: [:create]
+  before_update :set_last_updated_at
```
Comment (zdavis) [inline api/app/models/project.rb]: I think I understand what you're doing here. You're trying to avoid updating the update timestamp for a project when the draft mode changes. This approach feels wrong to me. None of this update timestamp logic feels like a proper concern for the Project model.

Rails 5 lets you not update a timestamp when you save the model by passing a touch: false option to the save method. For example: `project.save(touch: false)`

I think you should drop all this timestamp logic and just handle this in the project controller's update method. Currently, you have this:

`::Updaters::Project.new(project_params).update(@project)`

Why not change the `update` method in `Updaters` so it accepts a second options argument. In the controller, if the draft mode is the only thing being changed, you could pass `touch: false` to the update method, which could in turn pass the touch false option to the save method. This seems simpler and more reusable to me.

File: api/app/serializers/project_partial_serializer.rb [hunk 3]
```diff
@@ -7,7 +7,8 @@ class ProjectPartialSerializer < ActiveModel::Serializer
              :purchase_price_currency, :purchase_price, :purchase_version_label,
              :twitter_id, :instagram_id, :facebook_id, :hero_styles, :cover_styles,
              :avatar_styles, :recently_updated, :updated, :description_formatted, :slug,
-             :resource_kinds, :resource_tags, :avatar_color, :avatar_meta
+             :resource_kinds, :resource_tags, :avatar_color, :avatar_meta, :draft,
+             :last_updated_at
```
Comment (zdavis) [inline api/app/serializers/project_partial_serializer.rb]: Per above, I think this should be removed.

File: api/spec/models/project_spec.rb [hunk 4]
```diff
@@ -227,5 +237,31 @@
     end
   end
 
+  context "does not change the updated_at attribute" do
```
Comment (zdavis) [inline api/spec/models/project_spec.rb]: This test should be in the controller, not the model.

File: client/src/api/resources/favoriteProjects.js [hunk 5]
```diff
@@ -1,5 +1,7 @@
 export default {
-  index(filter = {}, page = {}) {
+  index(filterParams = {}, page = {}) {
+    const filter = filterParams;
+    filter.draft = false;
```
Comment (zdavis) [inline client/src/api/resources/favoriteProjects.js]: This seems weird, too... seems like if a project is draft, it shouldn't show up. The filter state should depend on whether the user is an admin or not.

File: client/src/api/resources/projects.js [hunk 6]
```diff
@@ -1,5 +1,7 @@
 export default {
-  index(filter = {}, page = {}) {
+  index(filterParams = {}, page = {}, includeDrafts = false) {
```
Comment (zdavis) [inline client/src/api/resources/projects.js]: This shouldn't be a client application concern at all, IMO. The API should decide what to return to the user — if the user is an admin, drafts are included in results. If they are not an admin, they can't see any drafts. If we handle it in the client rather than the server, we're not really hiding anything from the user.

File: client/src/api/resources/projects.js [hunk 7]
```diff
@@ -87,6 +89,7 @@ export default {
 
   featured(limit = 6, filterParams = {}) {
     const filter = filterParams;
+    filter.draft = false;
```
Comment (zdavis) [inline client/src/api/resources/projects.js]: This is incorrect as well. If a project is a draft project, it's never visible to a non-admin user, regardless of whether or not it's featured. See why we need to handle this check server side?

**Comment by zdavis** [comment]
This branch breaks backend project search. 
screen shot 2017-07-25 at 10 52 34 am

File: api/db/migrate/20170607190058_add_draft_to_projects.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    say_with_time 'Setting draft to FALSE for existing projects' do
+      Project.find_each do |project|
+        project.draft = false
+        project.save
+      end
+    end
+  end
+
+  def down
+    add_column :projects, :draft, :boolean
```
Comment (zdavis) [inline api/db/migrate/20170607190058_add_draft_to_projects.rb]: Shouldn't this be remove_column?

**Comment by zdavis** [comment]
I'm going to accept this. However, we need a way to indicate to the admin user which projects are in draft mode.

**Comment by naomiyaki** [comment]
Note that the draft label will be added to project detail heroes on the front-end as part of another issue to change the hero layout.

**Comment by zdavis** [comment]
@SMaxOwok or @naomiyaki We just need to remove the existing draft label:

image

If someone could do that, I'll accept it.

**Comment by naomiyaki** [comment]
@zdavis Max removed the label! Please give a second look and merge.

File: api/app/controllers/api/v1/projects_controller.rb [hunk 9]
```diff
@@ -27,6 +29,7 @@ def index
       def show
         @scope_for_projects = Project.friendly
         @project = load_project
+        authorize_action_for @project if @project.draft?
```
Comment (zdavis) [inline api/app/controllers/api/v1/projects_controller.rb]: Remove the condition — as above, the `if @project.draft?` is actually authorization logic that doesn't belong in this controller. Imagine if we has another controller that returned projects. We'd have to repeat this logic for every controller that touched projects. This PR is a good example of why DRY is about more than typing less and saving time — DRY code is more flexible and easier to refactor.

File: api/app/controllers/api/v1/projects_controller.rb [hunk 10]
```diff
@@ -37,7 +40,8 @@ def create
 
       def update
         @project = load_and_authorize_project
-        ::Updaters::Project.new(project_params).update(@project)
+        options = @project.draft? || project_params.dig(:draft) ? { touch: false } : {}
```
Comment (zdavis) [inline api/app/controllers/api/v1/projects_controller.rb]: Why is this line here? Again, this seems like the wrong place for this. Why not put it in the updater itself, if it's necessary?

Comment (zdavis) [inline api/app/controllers/api/v1/projects_controller.rb]: The logic here seems weird. "Any time a request to update a project comes in with a draft param, we don't change the timestamp on the project". What is this trying to accomplish? And why? And if this logic is important, why is it only in this controller?

File: api/app/services/demonstration/data_loader.rb [hunk 11]
```diff
@@ -29,6 +29,11 @@ def import_projects
         next if File.file?(File.join(child, ".skip"))
         Importer::Project.new(child, cli_user, @logger).import(true)
       end
+      # Make all demo projects live
```
Comment (zdavis) [inline api/app/services/demonstration/data_loader.rb]: I'd move this to Importer::Project and just have that service create non-draft projects by default. This block is problematic, too, because it's not just making all demo projects live—its making _all_ projects live. If we remove the part of the demo data loader that clears data before import, we'd have a problem.

File: api/app/services/importer/project.rb [hunk 12]
```diff
@@ -47,7 +47,7 @@ def upsert_project(include_texts)
       project.save if project.new_record?
       project.update(@project_json[:attributes])
       assign_project_attachments(project)
-      excludes = %w(cover_ avatar_ hero_)
+      excludes = %w(cover_ avatar_ hero_ draft)
```
Comment (zdavis) [inline api/app/services/importer/project.rb]: Why?

File: client/src/api/resources/favoriteProjects.js [hunk 13]
```diff
@@ -1,5 +1,6 @@
 export default {
-  index(filter = {}, page = {}) {
+  index(filterParams = {}, page = {}) {
```
Comment (zdavis) [inline client/src/api/resources/favoriteProjects.js]: What is the point of this change?

File: client/src/containers/backend/Dashboard.js [hunk 14]
```diff
@@ -19,7 +19,7 @@ const perPage = 5;
 export class DashboardContainer extends PureComponent {
   static fetchData = (getState, dispatch) => {
     const projectsRequest = request(
-      projectsAPI.index({}, { size: perPage }),
+      projectsAPI.index({}, { size: perPage }, true),
```
Comment (zdavis) [inline client/src/containers/backend/Dashboard.js]: Why this change? As far as I can tell, the method signature for the index method is: `index(filter = {}, page = {})`, yet you're passing a third argument to it.

File: client/src/containers/backend/Dashboard.js [hunk 15]
```diff
@@ -54,7 +54,7 @@ export class DashboardContainer extends PureComponent {
   updateResults(event = null, page = 1) {
     const pagination = { number: page, size: perPage };
     const action = request(
-      projectsAPI.index(this.state.filter, pagination),
+      projectsAPI.index(this.state.filter, pagination, true),
```
Comment (zdavis) [inline client/src/containers/backend/Dashboard.js]: See previous comment.

File: client/src/containers/frontend/Featured.js [hunk 16]
```diff
@@ -43,6 +43,13 @@ export class FeaturedContainer extends Component {
     subjects: PropTypes.array,
     dispatch: PropTypes.func
   };
+
```
Comment (zdavis) [inline client/src/containers/frontend/Featured.js]: Why is this revision part of this branch? How is it related to draft projects?

File: client/src/containers/frontend/Home.js [hunk 17]
```diff
@@ -99,6 +104,22 @@ export class HomeContainer extends Component {
     };
   }
 
+  fetchFilteredProjects = () => {
+    const filteredRequest = request(
+      projectsAPI.index(this.props.projectFilters),
```
Comment (zdavis) [inline client/src/containers/frontend/Home.js]: Shouldn't this be respecting pagination?

File: client/src/containers/frontend/Featured.js (hunk 16 above)
Comment (SMaxOwok) [inline client/src/containers/frontend/Featured.js]: This is necessary if you're on the /featured page and log in.  You won't see the draft projects unless we refetch.
